Filters API

AnastaZIuk opened the pull request:
status:

**done**:

- PNG Fix
- clamping 
- precomputed dithering
- blue noise dithering in denoiser tonemapper example
- gli swizzle feature
- compile time utilities
- swizzle in blit filter

AnastaZIuk committed:
fix flipping, make entry to clamping, add small assert in command line handler

AnastaZIuk committed:
Merge branch 'shader_pipeline' of https://github.com/buildaworldnet/IrrlichtBAW into denoiserFix

AnastaZIuk committed:
Merge pull request #12 from buildaworldnet/shader_pipeline

Shader pipeline

AnastaZIuk committed:
Merge branch 'denoiserFix' of https://github.com/AnastaZIuk/IrrlichtBAW into denoiserFix

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
 					constexpr auto MaxPlanes = 4;
 					const void* srcPix[MaxPlanes] = { commonExecuteData.inData+readBlockArrayOffset,nullptr,nullptr,nullptr };
 
+					auto clampedBuffer = getClampedSourcePixels(srcPix); // todo - compile time, do not waste memory
```
use `if constexpr` instead of a lambda, also clamp the pixels on output not input.

AnastaZIuk replied:
adjusted in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
+				// TESTING
+				auto getClampedSourcePixels = [&](auto sourcePixels)
+				{
+					constexpr auto MaxPlanes = 4;
```
is there a "plane count query" for `E_FORMAT` yet?

AnastaZIuk replied:
there is not as I look around

devshgraphicsprogramming replied:
make one ;)

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
+				auto getClampedSourcePixels = [&](auto sourcePixels)
+				{
+					constexpr auto MaxPlanes = 4;
+					auto buffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(blockDims.y * blockDims.x * inTexelOrBlockByteSize);
```
you just want a `double pixelVals[4];` instead of the buffer.

`decodePixels` takes an intrablock X,Y coordinate  to decode one pixel at a time.

AnastaZIuk replied:
adjusted in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
+							convertColor<Swizzle>(inFormat,outFormat, newSourcePixels,dstPix,blockX,blockY,state->swizzle);
 						else
-							convertColor<Swizzle>(inFormat,outFormat,srcPix,dstPix,blockX,blockY,*state);
+							convertColor<Swizzle>(inFormat,outFormat, newSourcePixels,dstPix,blockX,blockY,*state);
```
instead of having that weird lambda at the start, you want to break down the `convertColor<>` functions into their implementation constituent `decodePixels` and `encodePixels` that use a `uint64_t[4]` (sometimes casted to a `double[4]`) buffer for temporary storage.

Then sometimes (when you have the template parameter) depending on a `if constexpr` you want to shove in a `clamp` between the decode and encode.

devshgraphicsprogramming replied:
because you have the silly lambda you feel the need to allocate temporary storage for the block

devshgraphicsprogramming replied:
all my comments basically boil down to:
- see how `convertColor<>` works
- decompose it into a separate decode+encode
- put a clamp between the decode and the encode depending on the constexpr bool template parameter

AnastaZIuk replied:
adjusted in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
+							for (auto& value : decodeBuffer)
+								core::clamp<double, double>(value, 0, 1);
+
+							asset::encodePixelsRuntime(inFormat, reinterpret_cast<uint8_t*>(buffer->getPointer()) + (blockY * blockDims.x + blockDims.x) * inTexelOrBlockByteSize, decodeBuffer.data());
```
you're making an ungodly mess of it by decoding from INPUT_FORMAT ~~using the runtime function when we have compile time specializations which are used in `convertColor`~~ (actually you havent provided compile-time clamping at all for the compile-time format conversion from the general specialization of this class on lines 144-160), clamping, then converting to OUTPUT_FORMAT and saving to a buffer, then REINTERPRETING THE SAVED DATA as INPUT_FORMAT and having `convertColor` do the conversion to OUTPUT_FORMAT` AGAIN

AnastaZIuk replied:
it isn't valid anymore

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
+							asset::decodePixelsRuntime(inFormat, sourcePixels, decodeBuffer.data(), blockX, blockY);
+
+							for (auto& value : decodeBuffer)
+								core::clamp<double, double>(value, 0, 1);
```
clamp to output's format Minimum Value in the Range, and Maximum Value in the Range

NOT 0 and 1

P.S. also not all formats encode from `double`, so there's that to account for... see impl of `convertColor`

AnastaZIuk replied:
I was testing something, but ure right, adjusted in following commit (runtime)

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -188,20 +188,50 @@ class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CIma
+							std::array<double, MaxPlanes> decodeBuffer = {};
+							asset::decodePixelsRuntime(inFormat, sourcePixels, decodeBuffer.data(), blockX, blockY);
+
+							for (auto& value : decodeBuffer)
```
I'd do a loop from i=0 to output format channel count, if the output format is not `EF_UNDEFINED` (not runtime specified)

from i=0 to 0 otherwise

devshgraphicsprogramming replied:
for this class a `for (auto i=0; i<MaxChannels; i++)` is desired, for the compile-time general case
> 
> 
> you've only implemented the clamping for runtime (or at least 1-format-runtime, because we dont bother specializing for half-compile-time cases) format conversion.
> 
> I need an implementation for the both-formats-specified-compile-time case, which is the general class above, where `swizzle` lambda is defined between line 144 and 160

I want the for from 0 to output format's channel count (given by the template function, not runtime function)

AnastaZIuk replied:
adjusted for runtime in following commit, compile time left

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -164,8 +164,8 @@ class CSwizzleAndConvertImageFilter : public CImageFilter<CSwizzleAndConvertImag
-template<typename Swizzle>
-class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle> : public CImageFilter<CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle>>, public impl::CSwizzleAndConvertImageFilterBase<Swizzle>
+template<typename Swizzle, bool clamp>
+class CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle,clamp> : public CImageFilter<CSwizzleAndConvertImageFilter<EF_UNKNOWN,EF_UNKNOWN,Swizzle,clamp>>, public impl::CSwizzleAndConvertImageFilterBase<Swizzle>
```
you've only implemented the clamping for runtime (or at least 1-format-runtime, because we dont bother specializing for half-compile-time cases) format conversion.

I need an implementation for the both-formats-specified-compile-time case, which is the general class above, where `swizzle` lambda is defined between line 144 and 160

AnastaZIuk replied:
solved

AnastaZIuk committed:
make it work with runtime

AnastaZIuk committed:
make it work with compile time specialization

devshgraphicsprogramming commented on examples_tests/39.DenoiserTonemapper/main.cpp:
```diff
@@ -990,8 +990,8 @@ void main()
 			auto getConvertedPNGImageView = [&](core::smart_refctd_ptr<ICPUImage> image)
 			{
-				using CONVERSION_FILTER = CConvertFormatImageFilter<EF_UNKNOWN, EF_UNKNOWN, true>; // TESTING
 				constexpr auto pngFormat = EF_R8G8B8A8_SRGB;
```
drop the alpha, use R8G8B8_SRGB

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -126,22 +130,31 @@ class CSwizzleAndConvertImageFilter : public CImageFilter<CSwizzleAndConvertImag
 			auto perOutputRegion = [&blockDims,&state](const CMatchedSizeInOutImageFilterCommon::CommonExecuteData& commonExecuteData, CBasicImageFilterCommon::clip_region_functor_t& clip) -> bool
 			{
-				auto swizzle = [&commonExecuteData,&blockDims,&state](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos)
+				const auto clampBufferChannelsAmount = asset::getFormatChannelCount(outFormat);
```
is there a compile time version ? like `asset:getFormatChannelCount<E_FORMAT format>(void)` ?

devshgraphicsprogramming replied:
?

AnastaZIuk replied:
oh sorry, no there isn't

devshgraphicsprogramming replied:
make one.

AnastaZIuk replied:
roger that!

AnastaZIuk replied:
added and in use!

AnastaZIuk commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -151,10 +164,37 @@ class CSwizzleAndConvertImageFilter : public CImageFilter<CSwizzleAndConvertImag
+						if constexpr (clamp)
+							clampBuffer();
+							
+						asset::encodePixels<outFormat>(dstPix, reinterpret_cast<encodeBufferType*>(encodeBuffer));
```
as I think of it, shouldn't I move a pointer to be a `dstPix+= blockSize` each iteration in block loop?

AnastaZIuk replied:
okay, we do not support encoding to BC, it clears a little

devshgraphicsprogramming replied:
should still advance the pointer though XD ;)

AnastaZIuk replied:
yup, as well as I do it on SAT now

AnastaZIuk replied:
corrected in following commit

AnastaZIuk committed:
add compile-time version of asset:getFormatChannelCount, correct encode pointer

AnastaZIuk committed:
add some compile-time version of functions in EFormat.h

AnastaZIuk committed:
Merge branch 'shader_pipeline' of https://github.com/buildaworldnet/IrrlichtBAW into denoiserFix

AnastaZIuk committed:
add entry classes to dithering

AnastaZIuk commented on include/irr/asset/filters/dithering/CDither.h:
```diff
@@ -0,0 +1,47 @@
+						CState() {}
+						virtual ~CState() {}
+
+						ImageAsset imageAsset; // do we want to only handle ICPUImage or should I distinguish it to ICPUImageView?
```
we should make a decision about it

devshgraphicsprogramming replied:
dont hold this here, remember `CDither` is a base class for many things like:
- CWhiteNoiseDither
- CBayerMatrixDither
- CPrecomputedDither

AnastaZIuk replied:
not an issue now

AnastaZIuk commented on include/irr/asset/filters/dithering/IDither.h:
```diff
@@ -0,0 +1,36 @@
+
+                    struct ImageAsset
+                    {
+                        core::smart_refctd_ptr<asset::IAsset> asset;
```
it relates to the above https://github.com/buildaworldnet/IrrlichtBAW/pull/521#discussion_r475792929, I will code further if we decide

AnastaZIuk replied:
not sure if we use dds dithering images or whatever in future

devshgraphicsprogramming replied:
dont worry about this in the base class, only provide maximum extent that the dither must support.

AnastaZIuk replied:
okay, `IState` will only hold `TexelRange` now (we fill add more in future if needed)
`imageData` will be moved to precomputed dither class and will contain only strides and the buffer

devshgraphicsprogramming replied:
why should it have a member with texel range?

AnastaZIuk replied:
because we wanted to have max extent the dither support, so likely texel range I think, I can limit it to only extent

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CDither.h:
```diff
@@ -0,0 +1,47 @@
+			
+				float get(const state_type* state, const core::vectorSIMDu32& pixelCoord)
+				{
+					// use blue noise tile ability to fetch suitable texel pointer from texture
```
put a `return CRTP::get` here, this way its clear we expect the derived class to implement (also make function a static)

AnastaZIuk replied:
I did it wrong (compile time error), need to read probably more about CRTP

devshgraphicsprogramming replied:
you can't just do `CRTP::function` if `function` is not static

if its not static you need to do `static_cast<CRTP*>(this)->function`

AnastaZIuk replied:
done

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/IDither.h:
```diff
@@ -0,0 +1,36 @@
+        {
+            public:
+                virtual ~IDither() {}
+                virtual float pGet(const core::vectorSIMDu32& pixelCoord) = 0;
```
you need a pointer to `IState` as another parameter

AnastaZIuk replied:
solved

AnastaZIuk committed:
add dithering classes, provide a very little interface

AnastaZIuk commented on include/irr/asset/filters/dithering/CPrecomputedDither.h:
```diff
@@ -0,0 +1,71 @@
+				{
+					const auto& imageData = state->getImageData();
+
+					// if it has to be float, we should change something there cuz we aren't able to fetch the value due to channels
```
an issue!

devshgraphicsprogramming replied:
add a `int32_t channel parameter`

AnastaZIuk replied:
corrected

AnastaZIuk commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -150,11 +169,52 @@ class CSwizzleAndConvertImageFilter : public CImageFilter<CSwizzleAndConvertImag
+						*/
+
+						// too few info to do it - CHANNELS!
+						const float ditheredValue = blueNoise.dithering.pGet(&blueNoise.state, localOutPos + core::vectorSIMDu32(blockX, blockY));
```
an issue! https://github.com/buildaworldnet/IrrlichtBAW/pull/521/files#r476523336

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -150,11 +149,18 @@ class CSwizzleAndConvertImageFilter : public CImageFilter<CSwizzleAndConvertImag
+				using BLUE_NOISE_DITHER = asset::CPrecomputedDither;
+				BLUE_NOISE_DITHER dithering;
+				BLUE_NOISE_DITHER::state_type state;  /* how should I use asset manager there in constructor? i dont have an access */
+			} blueNoise;
```
make the dither type a template parameter and keep a `Dither*` pointer to the object in the state...

then obvisously you create the dither before you call the filter  ;)

AnastaZIuk replied:
corrected

devshgraphicsprogramming commented on include/irr/asset/IDither.h:
```diff
@@ -0,0 +1,52 @@
+                        };
+                };
+
+                virtual float pGet(const IState* state, const core::vectorSIMDu32& pixelCoord) = 0;
```
add an `int32_t channel` parameter,

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CBayerMatrixDither.h:
```diff
@@ -0,0 +1,36 @@
+				{
+					// TODO: to define in future
+				}
+		};
```
```
#if 0 // TODO
#endif
```
don't need it right now

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CDither.h:
```diff
@@ -5,40 +5,52 @@
+			- CWhiteNoiseDither
+			- CBayerMatrixDither
+			- CPrecomputedDither
+			- CHashDither
```
white noise is hash dither.

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CDither.h:
```diff
@@ -5,40 +5,52 @@
 						virtual ~CState() {}
 
-						ImageAsset imageAsset; // do we want to only handle ICPUImage or should I distinguish it to ICPUImageView?
+						ImageData imageData;
```
only precomputed keeps image data

AnastaZIuk replied:
corrected

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CDither.h:
```diff
@@ -5,40 +5,52 @@
 			private:
 			
-				float get(const state_type* state, const core::vectorSIMDu32& pixelCoord)
+				static float get(const state_type* state, const core::vectorSIMDu32& pixelCoord)
```
can be public

AnastaZIuk replied:
corrected

devshgraphicsprogramming commented on include/irr/asset/IDither.h:
```diff
@@ -0,0 +1,52 @@
+                    public:
+                        virtual ~IState() {}
+
+                        struct ImageData
```
no need for this here.

devshgraphicsprogramming replied:
goes to precomputed

AnastaZIuk replied:
corrected

AnastaZIuk committed:
nearly done, but I screwed up something probably - compile time errors, templates!

Something is wrong while putting DefaultSwizzle as template parameter - it says state is undefined (in IImageAssetHandlerBase.h)

AnastaZIuk committed:
undefined state with DefaultSizzle left, after the fix perform tests!

AnastaZIuk committed:
make it compile - beware of dependent names in future!

AnastaZIuk committed:
forgot about freeing memory, now need to perform dithering test

AnastaZIuk committed:
Make PrecomputedDither work - perform tests, add blue noise dithering to denoiser tonemapper example

devshgraphicsprogramming commented on examples_tests/39.DenoiserTonemapper/main.cpp:
```diff
@@ -14,6 +14,9 @@ using namespace irr;
 using namespace video;
 
+//! Comment if you want to apply blue noise on generated png files
+#define BLUE_NOISE_ON_PNG
```
assume always true ;)

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on examples_tests/39.DenoiserTonemapper/main.cpp:
```diff
@@ -1024,7 +1018,17 @@ void main()
 					
-					state.ditherState = _IRR_NEW(std::remove_pointer<decltype(state.ditherState)>::type, ditheringImage);
+					#ifdef BLUE_NOISE_ON_PNG
+					auto ditheringBundle = am->getAsset("../../media/blueNoiseDithering/LDR_RGBA.png", {});
```
load dither from EXR... you'll get better dither

AnastaZIuk replied:
not an issue anymore - discussed on discord

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CDither.h:
```diff
@@ -47,7 +47,7 @@ namespace irr
 				float get(const IDither::IState* state, const core::vectorSIMDu32& pixelCoord, const int32_t& channel)
 				{
-					return static_cast<CRTP*>(this)->get(reinterpret_cast<const typename CRTP::CState*>(state), pixelCoord, channel);
+					return static_cast<CRTP*>(this)->get(dynamic_cast<const typename CRTP::CState*>(state), pixelCoord, channel);
```
`static_cast`

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CPrecomputedDither.h:
```diff
@@ -52,12 +52,12 @@ namespace irr
 						} ditherImageData;
 
-						core::smart_refctd_ptr<asset::ICPUImage> ditheringImage;
+						const asset::ICPUImage* const ditheringImage;
```
you don't need to keep it around after you construct the state.

AnastaZIuk replied:
Look at how stare gets constructed, I think its okay

AnastaZIuk replied:
state

devshgraphicsprogramming replied:
you dont want to keep the image around (regions, etc.) just its contents.

AnastaZIuk replied:
corrected in following commit

AnastaZIuk committed:
Add swizzle feature to Blit Filter - NEEDS TESTING!

AnastaZIuk committed:
add dither state docs and missing dither img

AnastaZIuk committed:
Perform test swizzle on SAT - it works

AnastaZIuk committed:
change path to load LDR blue noise dithering

devshgraphicsprogramming commented on include/irr/asset/filters/CBlitImageFilter.h:
```diff
@@ -174,13 +174,26 @@ class CBlitImageFilter : public CImageFilter<CBlitImageFilter<KernelX,KernelX,Ke
 				KernelX					kernelX;
 				KernelY					kernelY;
 				KernelZ					kernelZ;
+				asset::DefaultSwizzle defaultSwizzle;
```
the swizzle type should be a template, because there are compile time swizzles and runtime polymorphic swizzles (see our other APIs)

devshgraphicsprogramming replied:
also you need an input swizzle and an output swizzle

AnastaZIuk replied:
I don't get, we do i need in swizzle and out swizzle if filter producent only one out image
How shoould i process in?

devshgraphicsprogramming replied:
Blit takes an input image and puts it in the output image, while applying convolution kernels (or other kernels)

If you can swizzle, then you should be able to swizzle before and after (after decode and before encode) the kernel.

devshgraphicsprogramming commented on include/irr/asset/filters/CBlitImageFilter.h:
```diff
@@ -416,6 +429,10 @@ class CBlitImageFilter : public CImageFilter<CBlitImageFilter<KernelX,KernelX,Ke
 
+								value_type swizzledSample[MaxChannels];
+								state->defaultSwizzle(sample, swizzledSample, MaxChannels);
+								memcpy(sample, swizzledSample, sizeof(value_type) * MaxChannels);
```
we don't need to memcpy, I think `decodePixels` has a variant that takes a swizzle.

AnastaZIuk replied:
You would not to force me to split what's inside decodepixels otherwise in convert swizzle  filter then

AnastaZIuk replied:
but i will check it

AnastaZIuk replied:
`decodePixels` doesn't have such variant - `convertColor` does, but we don't want it

AnastaZIuk replied:
not an issue anymore

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -150,11 +177,52 @@ class CSwizzleAndConvertImageFilter : public CImageFilter<CSwizzleAndConvertImag
+						uint64_t encodeBuffer[maxChannels] = {};
+
+						asset::decodePixels<inFormat>(srcPix, reinterpret_cast<decodeBufferType*>(decodeBuffer), blockX, blockY);
+						swizzle.template operator()<void, void>(decodeBuffer, encodeBuffer);
```
see how swizzle gets implemented?

AnastaZIuk replied:
?

devshgraphicsprogramming replied:
this is the "correct" way to implement Swizzle

AnastaZIuk replied:
okay

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CDither.h:
```diff
@@ -0,0 +1,56 @@
+						CState() {}
+						virtual ~CState() {}
+
+						TexelRange texelRange;
```
what is texel range?

AnastaZIuk replied:
i will change it to extent maybe?

devshgraphicsprogramming replied:
oh ok so its a measure of dimension?

AnastaZIuk replied:
It's extent + offset field

devshgraphicsprogramming replied:
ah kwl

devshgraphicsprogramming replied:
you only need extent to be honest... and also not sure what its doing in the CDither class

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CPrecomputedDither.h:
```diff
@@ -0,0 +1,89 @@
+					public:
+						CState(const asset::ICPUImage* const ditheringImage) 
+						{
+							ditherImageData.buffer = ditheringImage->getBuffer();
```
you can't just take over a buffer of an image (even if you could you should have constructed a smartrefctd ptr)

because the buffer might not be there or cover too small of an area.

You need the CopyImageToBuffer filter IIRC.

devshgraphicsprogramming replied:
or create a proxy image with its own buffer and flatten the input image into that.

AnastaZIuk replied:
what?

AnastaZIuk replied:
do we have such?

devshgraphicsprogramming replied:
https://github.com/buildaworldnet/IrrlichtBAW/blob/shader_pipeline/include/irr/asset/filters/CFlattenRegionsImageFilter.h

Should be a flatten & convert or you should just only accept images in the RGBA UNORM16 or UNORM32 formats

AnastaZIuk replied:
done, but not sure if I should still force RGBA UNORM16 or UNORM32 formats

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CPrecomputedDither.h:
```diff
@@ -0,0 +1,89 @@
+					double decodeBuffer[MAX_CHANNELS];
+					asset::decodePixelsRuntime(ditherImageData.format, srcPointers, decodeBuffer, 0, 0); // little slow
+
+					return static_cast<float>(decodeBuffer[channel]);
```
nope, the values should have already been decoded.

AnastaZIuk replied:
?

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -28,13 +27,17 @@ namespace impl
+			public:
+				Dither dither;
+				using DitherState = typename Dither::state_type;
+				DitherState* ditherState = nullptr;					//! To use dithering in the filter - allocate, create and make the dither state pointer valid
```
there should be some sort of conditional (`enable_if`) declaration of these member variables

Basically if I'd set `Dither=void` it means no dither decided at compile time.

devshgraphicsprogramming replied:
see how there's a separate state member for `PolymorphicSwizzle` and there isn't when the swizzle is not polymorphic

AnastaZIuk replied:
there is a IdentityDither class now according to DC conversation and it works fine (at compile time)

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleAndConvertImageFilter.h:
```diff
@@ -103,45 +117,58 @@ inline void DefaultSwizzle::operator()(const InT* in, OutT* out) const
+	Do a per-pixel recombination of image channels while converting
+*/
+
+template<E_FORMAT inFormat=EF_UNKNOWN, E_FORMAT outFormat=EF_UNKNOWN, typename Swizzle=DefaultSwizzle, bool Clamp = false, typename Dither = asset::CPrecomputedDither>
```
default dither `void` a.k.a. no dither.

AnastaZIuk replied:
done

AnastaZIuk committed:
Create CSwizzleableAndDitherableFilterBase.h, add some onDecode and onEncode functions, remove HDR precomputed img

AnastaZIuk committed:
Add flatten while filling CPrecomputedDither's state, add some docs

AnastaZIuk committed:
Make compile time swizzle convert filter work with new SwizzableAndDitherable API

AnastaZIuk committed:
Make entire CSwizzleAndConvertImageFilter work with half specialization as well. Need to figure out why we have access violation XD

devshgraphicsprogramming commented on examples_tests/39.DenoiserTonemapper/main.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <irrlicht.h>
 
 #include "CommandLineHandler.hpp"
+#include "../../include/irr/asset/filters/dithering/CPrecomputedDither.h"
```
this should already be included in `irr/asset/asset.h"

AnastaZIuk replied:
corrected

AnastaZIuk committed:
XD assert killed me XD

AnastaZIuk committed:
change CPrecomputedDither to hold decoded dithered image data - now `get` fetches decoded texel

AnastaZIuk committed:
TODO: perform white noise in CWhiteNoiseDither and change blit filter API to new one

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CWhiteNoiseDither.h:
```diff
@@ -21,20 +19,46 @@ namespace irr
+						}
 						virtual ~CState() {}
+
+						core::RandomSampler sampler;
```
nope, you dont want a random number generator to implement this... you want a hash function and feed it the `pixelCoord` and `channel` as input to permute into a single 32 bit value (that you then divide by ~0u to get a float in the correct range)

AnastaZIuk replied:
solved

AnastaZIuk committed:
Add Wang Hash to CWhiteNoiseDither - similar to white noise distribution hash function

AnastaZIuk committed:
Change CBlitImageFilter API, make it almost done (white noise isn't good enough, too high values on output somehow)

AnastaZIuk committed:
Resolve the issue with weird output, CBlitImageFilter now uses new API without any troubles

devshgraphicsprogramming commented on include/irr/asset/filters/CBlitImageFilter.h:
```diff
@@ -192,17 +197,15 @@ class CBlitImageFilter : public CImageFilter<CBlitImageFilter<KernelX,KernelX,Ke
 				KernelY					kernelY;
 				KernelZ					kernelZ;
-				asset::DefaultSwizzle defaultSwizzle;
 				bool normalize = false;
```
compile time template parameter would be better

AnastaZIuk replied:
everything is now templated

AnastaZIuk replied:
oh u mean `normalize`

AnastaZIuk replied:
so what kind of classes should I pass to `normalize`? or should it be only a bool in template param?

AnastaZIuk replied:
and it would do the same SAT's stuff

devshgraphicsprogramming replied:
> 
> 
> so what kind of classes should I pass to `normalize`? or should it be only a bool in template param?

SAT, ConvertSwizzle and Blit

do we have any more questions after discord chat?

AnastaZIuk replied:
no, there are not

AnastaZIuk replied:
corrected

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CPrecomputedDither.h:
```diff
@@ -58,11 +102,15 @@ namespace irr
 
 					private:
 
+						static constexpr auto decodeFormat = EF_R64G64B64A64_SFLOAT;
```
if the definition of `get` is `float` then your decode format is 32-bit float not 64bit ;)

AnastaZIuk replied:
Oh okay

devshgraphicsprogramming replied:
you want to use `floats` instead of `double`s basically.

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CPrecomputedDither.h:
```diff
@@ -26,30 +26,74 @@ namespace irr
 				class CState : public CDither::CState
 				{
 					public:
 						CState(const asset::ICPUImage* const ditheringImage) 
```
maybe take an `ICPUImageView` then one has a choice (indirectly) of the mip-map and layer that could be used for the dither.

AnastaZIuk replied:
Okay

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CWhiteNoiseDither.h:
```diff
@@ -29,12 +32,28 @@ namespace irr
+					};
+					
+					const auto hash = static_cast<double>(getWangHash());
+					return static_cast<float>(static_cast<double>(hash) / double(~0u));
```
your wang hash is done in 32bit uint, your return type is `float`... there's no need for using `double`s

AnastaZIuk replied:
Okay, so i will devide by float and stop using those casts

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CWhiteNoiseDither.h:
```diff
@@ -29,12 +32,28 @@ namespace irr
+					auto getWangHash = [&]()
+					{
+						IImage::SBufferCopy::getLocalByteOffset(pixelCoord, bufferStridesHash);
+						uint32_t seed = IImage::SBufferCopy::getLocalByteOffset(pixelCoord, bufferStridesHash) * channel;
```
dont use byte offsets, just compute a pixel index value `(((channel*extent.depth+z)*extent.height+y)*extent.width+x`

AnastaZIuk replied:
what about it https://github.com/buildaworldnet/IrrlichtBAW/pull/521#issuecomment-683958930?

AnastaZIuk replied:
corrected in following commit

AnastaZIuk committed:
Add normalize template parameter to the CSwizzleAndConvertImageFilter, create small interface for normalizing, change the seed in CWhiteNoiseDither

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleableAndDitherableFilterBase.h:
```diff
@@ -17,20 +17,124 @@ namespace irr
 		{
+			namespace detail
+			{
+				union NormalizeValues
```
you want a `struct` and then an "anonymous union" of the member variables

AnastaZIuk replied:
not an issue, discussed on the dc

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleableAndDitherableFilterBase.h:
```diff
@@ -17,20 +17,124 @@ namespace irr
+						*/
+
+						template<E_FORMAT format, typename Tenc>
+						void normalize(Tenc* encodeBuffer, const uint8_t& channels)
+						{
+							if constexpr (isSignedFormat<format>())
+								for (uint8_t channel = 0; channel < channels; ++channel)
+									encodeBuffer[channel] = (2.0 * encodeBuffer[channel] - oldMaxValue.f[channel] - oldMinValue.f[channel]) / (oldMaxValue.f[channel] - oldMinValue.f[channel]);
+							else
+								for (uint8_t channel = 0; channel < channels; ++channel)
+									encodeBuffer[channel] = (encodeBuffer[channel] - oldMinValue.f[channel]) / (oldMaxValue.f[channel] - oldMinValue.f[channel]);
+						}
+
+						/*
+							Runtime version of normalize
+
+							@see normalize
+						*/
+
+						template<typename Tenc>
+						void normalize(const E_FORMAT& format, Tenc* encodeBuffer, const uint8_t& channels)
+						{
```
`static_assert` and `assert` that the format is not a pure-integer format

AnastaZIuk replied:
corrected in following commit

devshgraphicsprogramming commented on include/irr/asset/filters/CSwizzleableAndDitherableFilterBase.h:
```diff
@@ -17,20 +17,124 @@ namespace irr
+						CNormalizeState() {}
+						virtual ~CNormalizeState() {}
+
+						NormalizeValues oldMaxValue, oldMinValue;
```
dont declare the members in this specialization

AnastaZIuk replied:
not an issue, discussed on the dc

devshgraphicsprogramming commented on include/irr/asset/filters/dithering/CDither.h:
```diff
@@ -47,7 +47,9 @@ namespace irr
 				{
-					return static_cast<CRTP*>(this)->get(static_cast<const typename CRTP::CState*>(state), pixelCoord, channel);
+					const auto& return_value = static_cast<CRTP*>(this)->get(static_cast<const typename CRTP::CState*>(state), pixelCoord, channel);
+					assert(return_value >= 0 && return_value <= 1, "The dither value must be limited to the range!");
```
only assert `#ifdef _IRR_DEBUG`

devshgraphicsprogramming replied:
also a non-static assert does not take a message argument AFAIK

AnastaZIuk replied:
corrected in following commit
